tests/cwd_test.py

```diff
@@ -0,0 +1,20 @@
+import os
+from subprocess import run, PIPE
+
+import pytest
+
+@pytest.mark.usefixtures('both_debug_modes', 'both_setsid_modes')
+def test_working_directories():
+    """The child process must start in the working directory in which
+    dumb-init was invoked, but dumb-init itself should not keep a
+    reference to that."""
+
+    # We need absolute path to dumb-init since we pass cwd=/tmp to get
+    # predictable output - so we can't rely on dumb-init being found
+    # in the "." directory.
+    proc = run((os.path.join(os.getcwd(), 'dumb-init'),
```

Comment:
I think this line fails under tox because dumb-init will be installed into the virtualenv's bin directory. You can see the failure in Travis:

```
=================================== FAILURES ===================================
________________________ test_working_directories[1-1] _________________________
    @pytest.mark.usefixtures('both_debug_modes', 'both_setsid_modes')
    def test_working_directories():
        """The child process must start in the working directory in which
        dumb-init was invoked, but dumb-init itself should not keep a
        reference to that."""
    
        # We need absolute path to dumb-init since we pass cwd=/tmp to get
        # predictable output - so we can't rely on dumb-init being found
        # in the "." directory.
        proc = run((os.path.join(os.getcwd(), 'dumb-init'),
                    'sh', '-c', 'readlink /proc/$PPID/cwd && readlink /proc/$$/cwd'),
>                  cwd="/tmp", stdout=PIPE, stderr=PIPE)
tests/cwd_test.py:17: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.7/subprocess.py:472: in run
    with Popen(*popenargs, **kwargs) as process:
/usr/lib/python3.7/subprocess.py:775: in __init__
    restore_signals, start_new_session)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <subprocess.Popen object at 0x7f25371729e8>
args = ['/test/dumb-init', 'sh', '-c', 'readlink /proc/$PPID/cwd && readlink /proc/$$/cwd']
executable = b'/test/dumb-init', preexec_fn = None, close_fds = True
pass_fds = (), cwd = '/tmp', env = None, startupinfo = None, creationflags = 0
shell = False, p2cread = -1, p2cwrite = -1, c2pread = 11, c2pwrite = 12
errread = 13, errwrite = 14, restore_signals = True, start_new_session = False
    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       restore_signals, start_new_session):
        """Execute program (POSIX version)"""
[...]
>               raise child_exception_type(errno_num, err_msg, err_filename)
E               FileNotFoundError: [Errno 2] No such file or directory: '/test/dumb-init': '/test/dumb-init'
```

(from https://travis-ci.org/github/Yelp/dumb-init/jobs/747386344)

Maybe you can use `shutil.which` to get the path instead? https://docs.python.org/3/library/shutil.html#shutil.which

Reply:
Hm, yes, with a grain of os.path.realpath added - shutil.which just returns `./dumb-init` when `.` is in path and dumb-init does exist in `.`. Let's try this one then.